private/buf/buflsp/mutex.go

```diff
@@ -0,0 +1,182 @@
+		mp.held = make(map[uint64]*mutex)
+	}
+
+	if isUnlock {
+		if held := mp.held[id]; held != mu {
+			panic(fmt.Sprintf("buflsp/mutex.go: attempted to unlock incorrect non-reentrant lock: %p -> %p", held, mu))
+		}
+
+		delete(mp.held, id)
+	} else {
+		if held := mp.held[id]; held != nil {
+			panic(fmt.Sprintf("buflsp/mutex.go: attempted to acquire two non-reentrant locks at once: %p -> %p", mu, held))
+		}
+
+		mp.held[id] = mu
+	}
+}
+
+// mutex is a sync.Mutex with some extra features.
+//
+// The main feature is reentrancy-checking. Within the LSP, we need to lock-protect many structures,
+// and it is very easy to deadlock if the same request tries to lock something multiple times.
+// To achieve this, Lock() takes a context, which must be modified by withRequestID().
+type mutex struct {
+	lock sync.Mutex
+	// This is the id of the context currently holding the lock.
+	who  atomic.Uint64
+	pool *mutexPool
+}
+
+// Lock attempts to acquire this mutex or blocks.
+//
+// Unlike [sync.Mutex.Lock], this takes a Context. If that context was updated with withRequestID,
+// this function will panic when attempting to lock the mutex while it is already held by a
+// goroutine using this same context.
+//
+// NOTE: to Lock() and Unlock() with the same context DO NOT synchronize with each other. For example,
+// attempting to lock this mutex from two different goroutines with the same context will
+// result in undefined behavior.
+//
+// Also unlike [sync.Mutex.Lock], it returns an idempotent unlocker function. This can be used like
+// defer mu.Lock()(). Note that only the outer function call is deferred: this is part of the
+// definition of defer. See https://go.dev/play/p/RJNKRcoQRo1. This unlocker can also be used to
+// defer unlocking but also unlock before the function returns.
+//
+// The returned unlocker is not thread-safe.
+func (mu *mutex) Lock(ctx context.Context) (unlocker func()) {
+	var unlocked bool
+	unlocker = func() {
+		if unlocked {
+			return
+		}
+		mu.Unlock(ctx)
+		unlocked = true
+	}
+
+	id := getRequestID(ctx)
+
+	if mu.who.Load() == id && id > 0 {
+		// We seem to have tried to lock this lock twice. Panic, and poison the lock.
+		mu.who.Store(poison)
+		panic("buflsp/mutex.go: non-reentrant lock locked twice by the same request")
+	}
+
+	mu.pool.check(id, mu, false)
+
+	// Ok, we're definitely not holding a lock, so we can block until we acquire the lock.
+	mu.lock.Lock()
+	mu.storeWho(id)
+
+	return unlocker
+}
+
+// Unlock releases this mutex.
+//
+// Unlock must be called with the same context that locked it, otherwise this function panics.
+func (mu *mutex) Unlock(ctx context.Context) {
+	id := getRequestID(ctx)
+	if mu.who.Load() != id {
+		panic("buflsp/mutex.go: lock was locked by one request and unlocked by another")
+	}
+
+	mu.storeWho(0)
+
+	mu.pool.check(id, mu, true)
+	mu.lock.Unlock()
+}
+
+func (mu *mutex) storeWho(id uint64) {
+	for {
+		// This has to be a CAS loop to avoid races with a poisoning p.
```

Comment:
Not really. If you are unable to swap it's confirmed to be poisoned as the lock is always held.
```go
if mu.who.CompareAndSwap(0, id) {
	panic("buflsp/mutex.go: non-reentrant lock locked twice by the same request")
}
```